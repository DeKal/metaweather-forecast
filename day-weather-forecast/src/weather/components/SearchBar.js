import React from 'react'
import clsx from 'clsx'
import PropTypes from 'prop-types'
import { CircularProgress, TextField } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import theme from 'core/styles/theme'
import Autocomplete from '@material-ui/lab/Autocomplete'
import InputAdornment from '@material-ui/core/InputAdornment'

import SearchIcon from '@material-ui/icons/Search'

const useStyles = makeStyles(() => ({
  container: {
    height: theme.spacing(6.75),
    borderRadius: theme.spacing(1)
  },
  inputBox: theme.typography.body1,
  helperText: theme.typography.body3,
  cssOutlinedInput: {
    borderRadius: theme.spacing(1),
    '&$cssFocused $notchedOutline': {
      borderColor: theme.colors.primary[1]
    },
    '&:hover:not($disabled):not($focused):not($error) $notchedOutline': {
      borderColor: theme.colors.primary[1]
    }
  },
  cssLabel: {
    '&.Mui-focused': {
      color: theme.colors.primary[1]
    }
  },
  disabled: {},
  focused: {},
  error: {},
  cssFocused: {},
  notchedOutline: {}
}))

export const matchQueryToListCity = (list, query) => {
  if (typeof query === 'string') {
    // In case suggested list has one item, just go for it
    if (list.length === 1) {
      return list[0]
    }

    // Find exact match
    const city = list.find(
      (item) => item.title.toLowerCase() === query.toLowerCase()
    )

    return city || null
  }
  return query
}

const SearchBar = ({
  className,
  label,
  required,
  placeholder,
  onChange,
  onChangeInput,
  error,
  maxLength,
  type,
  disabled,
  fullWidth,
  list,
  loading
}) => {
  const classes = useStyles()
  return (
    <Autocomplete
      freeSolo
      id="auto-complete-text-field"
      disableClearable
      options={list}
      onChange={(_, selectedValue) => {
        const city = matchQueryToListCity(list, selectedValue)
        onChange(city)
      }}
      getOptionLabel={(options) =>
        typeof options === 'string' ? options : options.title
      }
      renderInput={(params) => (
        <TextField
          type="text"
          label={label}
          className={clsx(classes.container, className)}
          variant="outlined"
          placeholder={placeholder}
          required={required}
          InputProps={{
            className: classes.inputBox,
            classes: !error
              ? {
                  root: classes.cssOutlinedInput,
                  focused: classes.cssFocused,
                  notchedOutline: classes.notchedOutline
                }
              : null,
            endAdornment: (
              <InputAdornment position="end">
                {loading ? (
                  <CircularProgress
                    size={25}
                    style={{ color: theme.colors.disabledText }}
                  />
                ) : (
                  <SearchIcon
                    color={error ? 'error' : 'inherit'}
                    htmlColor={theme.colors.disabledText}
                  />
                )}
              </InputAdornment>
            )
          }}
          InputLabelProps={{
            classes: {
              root: classes.cssLabel,
              focused: classes.cssFocused
            }
          }}
          inputProps={{
            ...params.inputProps,
            maxLength,
            type
          }}
          FormHelperTextProps={{
            className: classes.helperText
          }}
          error={error}
          disabled={disabled}
          fullWidth={fullWidth}
          ref={params.InputProps.ref}
          autoFocus
          onChange={(e) => {
            onChangeInput(e.target.value)
          }}
        />
      )}
    />
  )
}

export default SearchBar
SearchBar.propTypes = {
  label: PropTypes.string,
  error: PropTypes.bool,
  className: PropTypes.any,
  required: PropTypes.bool,
  placeholder: PropTypes.string,
  value: PropTypes.string,
  onChange: PropTypes.func,
  maxLength: PropTypes.number,
  type: PropTypes.string,
  disabled: PropTypes.bool,
  fullWidth: PropTypes.bool,
  list: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string,
      woeid: PropTypes.number
    })
  ),
  onChangeInput: PropTypes.func,
  loading: PropTypes.bool
}
SearchBar.defaultProps = {
  list: [],
  placeholder: '',
  label: 'Search',
  required: false,
  disabled: false,
  error: false,
  onChange: () => {},
  onChangeInput: () => {},
  fullWidth: false,
  loading: false
}
