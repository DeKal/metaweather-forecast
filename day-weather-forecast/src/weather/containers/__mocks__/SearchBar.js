/* eslint-disable react/prop-types */
import React from 'react'

const MockSearchBar = (props) => (
  <input
    onChange={(e) => {
      props.onChange({
        woeid: e.target.value
      })
      props.onChangeInput(e.target.value)
    }}
    data-testid="mock-search-bar"
    type="text"
  />
)

export default MockSearchBar
