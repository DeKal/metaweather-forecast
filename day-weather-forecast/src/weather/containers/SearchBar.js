import { connect } from 'react-redux'
import { getCityList, getLoadingStateForCityApi } from 'state/city/selectors'

import SearchBar from 'weather/components/SearchBar'

const mapStateToProps = (state) => ({
  list: getCityList(state),
  loading: getLoadingStateForCityApi(state)
})

export default connect(mapStateToProps)(SearchBar)
