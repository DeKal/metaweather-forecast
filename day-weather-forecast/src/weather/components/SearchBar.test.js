import React from 'react'
import { render } from '@testing-library/react'
import SearchBar, { matchQueryToListCity } from 'weather/components/SearchBar'

const list = [
  {
    title: 'London',
    woeid: 44418
  },
  {
    title: 'Lond',
    woeid: 44418
  },
  {
    title: 'Lo',
    woeid: 44418
  }
]

describe('SearchBar test', () => {
  test('render correctly SearchBar for normal case', () => {
    const { container } = render(
      <SearchBar placeholder="Search" label="label" list={list} />
    )
    expect(container).toMatchSnapshot()
  })

  test('render correctly SearchBar for error case', () => {
    const { container } = render(
      <SearchBar placeholder="Search" label="label" error />
    )
    expect(container).toMatchSnapshot()
  })

  test('render correctly SearchBar for disabled case', () => {
    const { container } = render(
      <SearchBar placeholder="Search" label="label" disabled />
    )
    expect(container).toMatchSnapshot()
  })

  describe('matchQueryToListCity test', () => {
    test('given a string, should return a match city', () => {
      const city = matchQueryToListCity(list, 'lond')
      const wanted = {
        title: 'Lond',
        woeid: 44418
      }
      expect(city).toEqual(wanted)
    })

    test('given list only one city, should return a match city', () => {
      const listWithOneCity = [
        {
          title: 'London',
          woeid: 44418
        }
      ]
      const city = matchQueryToListCity(listWithOneCity, 'lond')
      const wanted = {
        title: 'London',
        woeid: 44418
      }
      expect(city).toEqual(wanted)
    })

    test('given an unmatched string, should return null in case not found', () => {
      const city = matchQueryToListCity(list, 'loz')
      const wanted = null
      expect(city).toEqual(wanted)
    })

    test('given a query, should return back that query', () => {
      const city = matchQueryToListCity(list, {
        title: 'Lo',
        woeid: 44418
      })
      const wanted = {
        title: 'Lo',
        woeid: 44418
      }
      expect(city).toEqual(wanted)
    })
  })
})
