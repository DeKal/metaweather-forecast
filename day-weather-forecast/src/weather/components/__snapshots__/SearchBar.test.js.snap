// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar test render correctly SearchBar for disabled case 1`] = `
<div>
  <div
    aria-expanded="false"
    class="MuiAutocomplete-root"
    role="combobox"
  >
    <div
      class="MuiFormControl-root MuiTextField-root makeStyles-container-29"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root makeStyles-cssLabel-33 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-disabled Mui-disabled"
        data-shrink="false"
      >
        label
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root makeStyles-cssOutlinedInput-32 makeStyles-inputBox-30 Mui-disabled Mui-disabled MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
      >
        <input
          aria-autocomplete="list"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused Mui-disabled Mui-disabled MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
          id="auto-complete-text-field"
          placeholder="Search"
          spellcheck="false"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            color="#9E9E9E"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </div>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-39 MuiOutlinedInput-notchedOutline makeStyles-notchedOutline-38"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-41"
          >
            <span>
              label
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchBar test render correctly SearchBar for error case 1`] = `
<div>
  <div
    aria-expanded="false"
    class="MuiAutocomplete-root Mui-focused"
    role="combobox"
  >
    <div
      class="MuiFormControl-root MuiTextField-root makeStyles-container-15"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root makeStyles-cssLabel-19 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-error Mui-error Mui-focused Mui-focused makeStyles-cssFocused-23"
        data-shrink="true"
      >
        label
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root makeStyles-inputBox-16 Mui-error Mui-error Mui-focused Mui-focused MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
      >
        <input
          aria-autocomplete="list"
          aria-invalid="true"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
          id="auto-complete-text-field"
          placeholder="Search"
          spellcheck="false"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorError"
            color="#9E9E9E"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </div>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-25 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-27 PrivateNotchedOutline-legendNotched-28"
          >
            <span>
              label
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchBar test render correctly SearchBar for normal case 1`] = `
<div>
  <div
    aria-expanded="false"
    class="MuiAutocomplete-root Mui-focused"
    role="combobox"
  >
    <div
      class="MuiFormControl-root MuiTextField-root makeStyles-container-1"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root makeStyles-cssLabel-5 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-focused Mui-focused makeStyles-cssFocused-9"
        data-shrink="true"
      >
        label
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root makeStyles-cssOutlinedInput-4 makeStyles-inputBox-2 Mui-focused Mui-focused makeStyles-cssFocused-9 MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
      >
        <input
          aria-autocomplete="list"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
          id="auto-complete-text-field"
          placeholder="Search"
          spellcheck="false"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            color="#9E9E9E"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </div>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-11 MuiOutlinedInput-notchedOutline makeStyles-notchedOutline-10"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-13 PrivateNotchedOutline-legendNotched-14"
          >
            <span>
              label
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
